text/0219-ecs-service-extensions.md

```diff
@@ -0,0 +1,503 @@
+service which is agnostic to whether it is deployed on EC2 or Fargate.
+
+## Tenants
+
+To solve the problems with ECS construct implemention we propose adding a new `Service`
+class with the following goals:
+
+- __One service class to learn__: Instead of a collection of many different L3 classes
+  that all have different names there is a single class which can implement any number
+  of different forms depending on what the customer wants. Customers don't need to learn
+  the names for different L3 constructs. All ECS architectures start the same way: with
+  the same `Service` class.
+- __Easy to extend and customize__: It is easy to extend the `Service` class with new
+  behaviors and features, in any combination required. The `Service` class comes with
+  batteries included: a set of easy to use extensions created and maintained by the AWS team.
+  However there are also clear instructions and examples of how customers can create their
+  own custom extensions. Third parties will be enabled to also create and distribute
+  extensions for their own services, for example a provider like Datadog could easily
+  build an extension that automatically adds Datadog agent to an ECS service.
+- __Extensions self configure and self provision__: Service extensions provision their
+  own resources as well as configuring the ECS service to use those resources. For example
+  a load balancer extension creates its own load balancer and also attaches the ECS service
+  to the load balancer. An App Mesh extension creates its own App Mesh virtual service and
+  virtual node as well as configuring the ECS service and ECS task definition to have the
+  right Envoy sidecar and proxy settings.
+- __Extensions are aware of each other__: There are no hidden "gotchas" when enabling a
+  service extension. Extensions know of each other's existence. For example the application's
+  container is aware that there is an Envoy proxy and that it must wait until the Envoy
+  proxy has started before the app starts.
+- __Service class is agnostic to EC2 vs Fargate__: Rather than choosing the capacity strategy
+  in the class name customers just create a service, and add it to an environment. The
+  service automatically modifies its settings as needed to adjust to the capacity available
+  in the environment.
+
+# Design Summary
+
+In this proposal we introduce two new constructs:
+
+- `Environment` - A simple construct which wraps up a VPC, cluster, and other top level
+  resources needed for ECS services to run on an AWS account
+- `Service` - A single application that runs inside an `Environment`
+
+And two new associated foundational classes used by the `Service` construct:
+
+- `ServiceDescription` - A class which is utilized to build out a description of the
+  application and the features that should be attached to the application. This description
+  is consumed by the `Service` class
```

Comment:
In the CDK, everything is a "description". What the value in decoupling "Service" and "ServiceDescription"?

Reply:
I actually in the PR have an alternate implementation which abuses the `construct.prepare()` method to allow lazy construction of the service's L2 constructs at synthesis time, you can see how that works here: https://github.com/aws/aws-cdk/blob/e60b7fdff95bae6d3e1980a64c998dc07d3efbb8/packages/%40aws-cdk/aws-ecs-patterns/test/service-addons/integ.all-service-addons.ts

While the API is simpler and does not require a `ServiceDescription` it leads to unexpected behavior as the underlying L2 constructs do not exist until synthesis time, so they can not be referenced in your stacks, you can't get attributes off them, etc. In discussions with the team we agreed that a `ServiceDescription` that is passed to the `Service` better fit the traditional model of L3 constructs, where all the underlying L2 constructs were being created in the `Service` constructor and are immediately available and immutable from that point forward